### PR TITLE
feat: [sc-97] Celebrate for backend

### PR DESCRIPTION
**Pull request by gubbih**
Story details: https://app.shortcut.com/5-semester/story/97
beff

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -18,6 +19,8 @@ app.get('/', (req: Request, res: Response) => {
     'Hello dumbass, this is the backend. Go to <a href="https://localhost:3010/">localhost:3010/</a> to see the app',
   );
 });
+//something something
```
Comment (mjrj97) [inline src/index.ts:22]:
```suggestion
// Catch celebrate validation errors
```

File: src/jest/events.test.ts [hunk 2]
```diff
@@ -89,54 +89,63 @@ describe('FIREBASE Event tests', () => {
...
 
   test('Get one event', async () => {
-    const res = await request(app).get('/events/' + testEvent2.orgId + '/' + testEvent2.id);
-    expect(res.body!.title).toBe(testEvent2.title);
-    expect(res.body?.description).toBe(testEvent2.description);
-    expect(res.body!.start).toStrictEqual(testEvent2.start.toISOString());
-    expect(res.body!.end).toStrictEqual(testEvent2.end.toISOString());
-    expect(res.body!.orgId).toBe(testEvent2.orgId);
+    const res = await request(app).get('/events/' + orgId + '/' + expressId);
+    expect(res.body!.title).toBe(expressEvent.title);
```
Comment (mjrj97) [inline src/jest/events.test.ts:116]:
```suggestion
    expect(res.statusCode).toBe(200);
    expect(res.body!.title).toBe(expressEvent.title);
```

File: src/jest/events.test.ts [hunk 3]
```diff
@@ -89,54 +89,63 @@ describe('FIREBASE Event tests', () => {
...
-    const corretId = res.body.find((ev: Event) => ev.id === testEvent2.id);
-    expect(corretId!.title).toBe(testEvent2.title);
-    expect(corretId?.description).toBe(testEvent2.description);
-    expect(corretId!.start).toStrictEqual(testEvent2.start.toISOString());
-    expect(corretId!.end).toStrictEqual(testEvent2.end.toISOString());
-    expect(corretId!.orgId).toBe(testEvent2.orgId);
-    expect(res.body.length).toBe(2);
+    const res = await request(app).get('/events/' + orgId);
+    const corretId = res.body.find((ev: Event) => ev.id === expressId);
+    expect(corretId!.title).toBe(expressEvent.title);
```
Comment (mjrj97) [inline src/jest/events.test.ts:126]:
```suggestion
    expect(res.statusCode).toBe(200);
    expect(corretId!.title).toBe(expressEvent.title);
```

File: src/routes/users.ts [hunk 4]
```diff
@@ -14,21 +14,43 @@ import {
...
+/*******************/
 
-// create user
 router.post(
   '/:orgId/',
-  userValidationRules,
+  celebrate({
+    [Segments.BODY]: Joi.object().keys({
+      firstName: Joi.string()
+        .alphanum()
```
Comment (mjrj97) [inline src/routes/users.ts:31]: Numbers should not be allowed

```suggestion
```

File: src/routes/users.ts [hunk 5]
```diff
@@ -14,21 +14,43 @@ import {
...
+        .regex(/^[a-zæøåA-ZÆØÅ\\-\s]+$/)
+        .min(2)
+        .max(64)
+        .required(),
+      lastName: Joi.string()
+        .alphanum()
```
Comment (mjrj97) [inline src/routes/users.ts:37]: Numbers should not be allowed
```suggestion
```

File: src/routes/users.ts [hunk 6]
```diff
@@ -14,21 +14,43 @@ import {
...
+        .regex(/^[a-zæøåA-ZÆØÅ\\-\s]+$/)
+        .min(2)
+        .max(64)
+        .required(),
+      email: Joi.string().required().email(),
+      password: Joi.string()
+        .required()
+        .regex(/^(?=.*\d)(?=.*[!@#$%^&*])(?=.*[a-zæøå)(?=.*[A-ZÆØÅ]).{8,}$/)
+        .max(64),
+      repeat_password: Joi.ref('password'),
```
Comment (mjrj97) [inline src/routes/users.ts:47]: 🔥 good thought

File: src/routes/users.ts [hunk 7]
```diff
@@ -37,16 +59,25 @@ router.post(
...
-//user logs in
 router.post(
   `/:orgId/login`,
-  userLoginValidationRules,
+  celebrate({
+    [Segments.BODY]: Joi.object().keys({
+      email: Joi.string().required().email(),
+      password: Joi.string()
+        .required()
+        .regex(/^(?=.*\d)(?=.*[!@#$%^&*])(?=.*[a-zæøå)(?=.*[A-ZÆØÅ]).{8,}$/)
```
Comment (mjrj97) [inline src/routes/users.ts:77]: What even is this 💀 hope it works

File: src/routes/users.ts [hunk 8]
```diff
@@ -37,16 +59,25 @@ router.post(
...
+        .max(64),
```
Comment (mjrj97) [inline src/routes/users.ts:78]:
```suggestion
        .min(8)
        .max(64),
```

File: src/routes/users.ts [hunk 9]
```diff
@@ -14,21 +14,43 @@ import {
...
+        .alphanum()
+        .regex(/^[a-zæøåA-ZÆØÅ\\-\s]+$/)
+        .min(2)
+        .max(64)
+        .required(),
+      email: Joi.string().required().email(),
+      password: Joi.string()
+        .required()
+        .regex(/^(?=.*\d)(?=.*[!@#$%^&*])(?=.*[a-zæøå)(?=.*[A-ZÆØÅ]).{8,}$/)
+        .max(64),
```
Comment (mjrj97) [inline src/routes/users.ts:46]:
```suggestion
        .min(8)
        .max(64),
```

**Review by mjrj97 — CHANGES_REQUESTED** [review]
Good job man. Celebrate 🥳

**Review by mjrj97 — APPROVED** [review]
Good job my guy 🎖️

File: src/routes/users.ts [hunk 10]
```diff
@@ -14,21 +14,42 @@ import {
...
+        .min(2)
+        .max(64)
+        .required(),
+      email: Joi.string().required().email(),
+      password: Joi.string()
+        .required()
+        .regex(/^(?=.*\d)(?=.*[!@#$%^&*])(?=.*[a-zæøå)(?=.*[A-ZÆØÅ]).{8,}$/)
+        .min(8)
+        .max(64),
+      repeat_password: Joi.ref('password'),
```
Comment (mjrj97) [inline src/routes/users.ts:46]:
```suggestion
      repeatPassword: Joi.ref('password'),
```

**Review by mjrj97 — CHANGES_REQUESTED** [review]
camelCase

File: src/jest/users.test.ts [hunk 11]
```diff
@@ -76,3 +68,91 @@ describe('FIREBASE User tests', () => {
...
+      .send(user1);
+
+    expressId = res.body.id;
+    expressToken = res.body.token;
+    hashedPassword = res.body.password;
+    expect(res.statusCode).toBe(201);
+    expect(res.body.firstName).toBe(user1.firstName);
+    expect(res.body.lastName).toBe(user1.lastName);
+    expect(res.body.email).toBe(user1.email);
+    expect(res.body.password).toBeDefined();
```
Comment (mjrj97) [inline src/jest/users.test.ts:99]: Passwords should NEVER be returned by the API. Not hashed or plain.
```suggestion
    expect(res.body).not.toHaveProperty('password');
```

File: src/jest/users.test.ts [hunk 12]
```diff
@@ -76,3 +68,91 @@ describe('FIREBASE User tests', () => {
...
+      .send({
+        email: user1.email,
+        password: user1.password,
+      })
+      .set('Authorization', 'Bearer ' + expressToken);
+    expect(res.statusCode).toBe(200);
+    expect(res.body.user.firstName).toBe(user1.firstName);
+    expect(res.body.user.lastName).toBe(user1.lastName);
+    expect(res.body.user.email).toBe(user1.email);
+    expect(res.body.user.password).not.toBe(hashedPassword);
```
Comment (mjrj97) [inline src/jest/users.test.ts:113]:
```suggestion
    expect(res.body).not.toHaveProperty('password');
```

File: src/jest/users.test.ts [hunk 13]
```diff
@@ -76,3 +68,91 @@ describe('FIREBASE User tests', () => {
...
+    expect(user.email).toBe(user1.email);
+  });
+  test('Get user by token', async () => {
+    const res = await request(app)
+      .get('/users/' + orgid + '/me')
+      .set('Authorization', 'Bearer ' + expressToken);
+    expect(res.statusCode).toBe(200);
+    expect(res.body.firstName).toBe(user1.firstName);
+    expect(res.body.lastName).toBe(user1.lastName);
+    expect(res.body.email).toBe(user1.email);
```
Comment (mjrj97) [inline src/jest/users.test.ts:132]:
```suggestion
    expect(res.body.email).toBe(user1.email);
    expect(res.body).not.toHaveProperty('password');
```

File: src/jest/users.test.ts [hunk 14]
```diff
@@ -76,3 +68,91 @@ describe('FIREBASE User tests', () => {
...
 });
+
+/******************/
+/**** Express  ****/
+/******************/
+
+describe('EXPRESS user routes', () => {
+  let expressId: string;
+  let expressToken: string;
+  let hashedPassword: string;
```
Comment (mjrj97) [inline src/jest/users.test.ts:79]:
```suggestion
```

File: src/jest/users.test.ts [hunk 15]
```diff
@@ -76,3 +68,91 @@ describe('FIREBASE User tests', () => {
...
+    password: 'Test123!!!',
+  };
+  test('Create user', async () => {
+    const res = await request(app)
+      .post('/users/' + orgid)
+      .send(user1);
+
+    expressId = res.body.id;
+    expressToken = res.body.token;
+    hashedPassword = res.body.password;
```
Comment (mjrj97) [inline src/jest/users.test.ts:94]:
```suggestion
```

File: src/jest/users.test.ts [hunk 16]
```diff
@@ -76,3 +68,91 @@ describe('FIREBASE User tests', () => {
...
+    };
+    const res = await request(app)
+      .put('/users/' + orgid + '/' + expressId)
+      .set('Authorization', 'Bearer ' + expressToken)
+      .send(user2);
+    expect(res.statusCode).toBe(200);
+    expect(res.body.firstName).toBe(user2.firstName);
+    expect(res.body.lastName).toBe(user2.lastName);
+    expect(res.body.email).toBe(user2.email);
+    expect(res.body.password).not.toBe(hashedPassword);
```
Comment (mjrj97) [inline src/jest/users.test.ts:149]:
```suggestion
    expect(res.body).not.toHaveProperty('password');
```

File: src/jest/users.test.ts [hunk 17]
```diff
@@ -76,3 +68,91 @@ describe('FIREBASE User tests', () => {
...
+  });
```
Comment (mjrj97) [inline src/jest/users.test.ts:150]: Maybe also make a login after this request, to see if the password has been changed.

File: src/jest/users.test.ts [hunk 18]
```diff
@@ -76,3 +68,91 @@ describe('FIREBASE User tests', () => {
...
+  test('Delete user', async () => {
+    const res = await request(app)
+      .delete('/users/' + orgid + '/' + expressId)
+      .set('Authorization', 'Bearer ' + expressToken);
+    console.log('logind user body thing', res.body);
```
Comment (mjrj97) [inline src/jest/users.test.ts:155]:
```suggestion
```

**Review by mjrj97 — CHANGES_REQUESTED** [review]

File: src/jest/users.test.ts (hunk 17 above)
Comment (gubbih) [reply src/jest/users.test.ts:164]: done thanks for the idea <3

**Review by mjrj97 — APPROVED** [review]
Proud of you champ

**Pull request merged**